### PR TITLE
Dev histogram

### PR DESCRIPTION
**Pull request by lotoussa**
Can still be improve but i want to mr and then start new branch for scatter_plot to work on everything.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -134,5 +134,3 @@ dmypy.json
 # Generated PDF
 *.pdf
 
-# numpy files data
-*.npy
```
Comment (xaviave) [inline .gitignore]: ?

File: sources/Tools/Visualiser.py [hunk 2]
```diff
@@ -62,13 +62,14 @@ def _add_exclusive_args(self, parser):
     """
 
     @staticmethod
-    def _save_as_pdf(figures):
+    def _save_as_pdf(file_name="data_visualizer", figures=None):
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: figures is not an optionnal parameter and file_name either (i didn;t change it)

File: sources/Tools/DatasetHandler.py [hunk 3]
```diff
@@ -113,10 +118,17 @@ def __init__(self, parse: bool = False, train: bool = False):
...
     """
 
-    def save_header(self, header_file: str = default_header_file):
+    def save_header(self, header_file: str = default_header_files["advanced"]):
         self._save_npy(header_file, self.analysed_header)
 
-    def load_header(self, header_file: str = default_header_file):
+    def load_header(self):
+        # test if not crash with all program
+        header_file = self.default_header_files[self.header_visualizer]
```
Comment (xaviave) [inline sources/Tools/DatasetHandler.py]: Big mistake, header visualizer don't exist in this class

File: sources/Tools/DatasetHandler.py [hunk 4]
```diff
@@ -113,10 +118,17 @@ def __init__(self, parse: bool = False, train: bool = False):
         Public methods
     """
 
-    def save_header(self, header_file: str = default_header_file):
+    def save_header(self, header_file: str = default_header_files["advanced"]):
         self._save_npy(header_file, self.analysed_header)
 
-    def load_header(self, header_file: str = default_header_file):
+    def load_header(self):
```
Comment (xaviave) [inline sources/Tools/DatasetHandler.py]: why did you delete the header_file parameter?

File: sources/Tools/DatasetHandler.py [hunk 5]
```diff
@@ -113,10 +118,17 @@ def __init__(self, parse: bool = False, train: bool = False):
...
+        # test if not crash with all program
+        header_file = self.default_header_files[self.header_visualizer]
+        print(self.header_visualizer)
+        print(self.func_visualizer)
```
Comment (xaviave) [inline sources/Tools/DatasetHandler.py]: delete the print please

File: sources/Tools/Visualiser.py [hunk 6]
```diff
@@ -196,17 +200,25 @@ def _exit(exception=None, message="Error", mod=-1):
...
+            self.header_visualizer = k
+            self.func_visualizer = v
 
     """
         Public methods
     """
 
     def visualizer(self, header):
-        matplotlib.use("pdf")
+        # matplotlib.use("pdf")
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: the allow the pdf backend, why did you comment it?

File: sources/Tools/Visualiser.py [hunk 7]
```diff
@@ -196,17 +200,25 @@ def _exit(exception=None, message="Error", mod=-1):
         logging.error(f"{message}")
         sys.exit(mod)
 
+    @property
+    def _get_houses(self):
+        return list(set(self.raw_data.loc[:, "Hogwarts House"]))
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: check numpy.unique()

File: sources/Tools/Visualiser.py [hunk 8]
```diff
@@ -196,17 +200,25 @@ def _exit(exception=None, message="Error", mod=-1):
...
+    @property
+    def _get_houses(self):
+        return list(set(self.raw_data.loc[:, "Hogwarts House"]))
+
     def __init__(self, func=_scatter_plot_visualizer):
         # could add different matplotlib backend | for now to much work
         super().__init__()
-        self.visualizer_func = self.get_args("type_visualizer", default_value=func)
+        # check this setter for every program usable to prevent crash
+        for k, v in self.get_args("type_visualizer", default_value={"advanced": func}).items():
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: i dont really like the fact that the for loop is just a for one initialisation
try more a type_visualiser={"name": "foo", "func": _bar}

File: sources/Tools/Visualiser.py [hunk 9]
```diff
@@ -62,13 +62,14 @@ def _add_exclusive_args(self, parser):
...
+    def _save_as_pdf(file_name="data_visualizer", figures=None):
         try:
-            with PdfPages(f"data_visualizer{len(figures)}.pdf") as pdf:
+            file_name = f"{file_name}_{len(figures)}.pdf"
+            with PdfPages(file_name) as pdf:
                 for f in figures:
                     pdf.savefig(f, bbox_inches="tight")
         except Exception as e:
-            logging.error(f"{e}\nError while creating data_visualizer{len(figures)}.pdf")
+            logging.error(f"{e}\nError while creating data_visualizer.pdf")
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: _exit func?

File: sources/Tools/Visualiser.py [hunk 10]
```diff
@@ -126,25 +127,28 @@ def _process_bar_data(raw_data, head):
...
-        ax.set_xticks(x)
-        ax.set_xticklabels(houses)
-        ax.legend()
-        self._auto_label(ax, rects1)
-        self._auto_label(ax, rects2)
-        fig.tight_layout()
-        return fig
+    @staticmethod
+    def _create_histogram(
+        feature, raw_data, filters, picker="Hogwarts House", xlabel="Marks", ylabel="Students"
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: not generic

File: sources/histogram.py [hunk 11]
```diff
@@ -2,5 +2,8 @@
 
 # Quel cours de Poudlard a une repartition des notes homogenes entre les quatres maisons ?
 
+# think about "mugging" ? google.ml.doc
```
Comment (xaviave) [inline sources/histogram.py]: i don't watch this type of porn

File: sources/Tools/Visualiser.py [hunk 12]
```diff
@@ -196,17 +200,25 @@ def _exit(exception=None, message="Error", mod=-1):
...
         Public methods
     """
 
     def visualizer(self, header):
-        matplotlib.use("pdf")
+        # matplotlib.use("pdf")
         if self.raw_data.empty:
             self._exit(message="Please init raw_data")
-        self.visualizer_func(header)
+        self.houses = self._get_houses
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: @proprety is not set to be use like this

File: sources/Tools/Visualiser.py [hunk 13]
```diff
@@ -126,25 +127,28 @@ def _process_bar_data(raw_data, head):
...
+            plt.xlim(mn, mx)
+            kde_xs = np.linspace(mn, mx, 301)
+            kde = st.gaussian_kde(data)
+            plt.plot(kde_xs, kde.pdf(kde_xs), label=elem)
+            plt.title(f"{elem} {feature}")
+        plt.xlabel(xlabel)
+        plt.ylabel(ylabel)
+        plt.yticks([])
+        plt.legend(filters, loc="upper left")
+        plt.show()
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: don't show it please, save it or show it at the end of the program

**Review by xaviave — CHANGES_REQUESTED** [review]
Please clean all the function, a logging.warning is not set to be in and date_visualizer is not use too

File: sources/Tools/DatasetHandler.py (hunk 3 above)
Comment (lotoussa) [reply sources/Tools/DatasetHandler.py]: It's initialised in Visualiser before entering in load_header

File: sources/Tools/DatasetHandler.py (hunk 4 above)
Comment (lotoussa) [reply sources/Tools/DatasetHandler.py]: Cause the arg wasn't used anywhere in the code by the called function and the new way to access to it is more efficient, it depend on the self.args content.

File: sources/Tools/DatasetHandler.py (hunk 5 above)
Comment (lotoussa) [reply sources/Tools/DatasetHandler.py]: I was using it to do scatter plot, i can remove now

File: sources/histogram.py (hunk 11 above)
Comment (lotoussa) [reply sources/histogram.py]: this is just to remind me of something, i can delete.

File: sources/Tools/Visualiser.py (hunk 12 above)
Comment (lotoussa) [reply sources/Tools/Visualiser.py]: i've apply some change on it.

File: sources/Tools/Visualiser.py (hunk 13 above)
Comment (lotoussa) [reply sources/Tools/Visualiser.py]: Sure, it was just the time to test the all plot program before rendering it in pdf files

File: sources/Tools/Visualiser.py (hunk 6 above)
Comment (lotoussa) [reply sources/Tools/Visualiser.py]: the time to test with plt.show()

File: sources/Tools/Visualiser.py (hunk 7 above)
Comment (lotoussa) [reply sources/Tools/Visualiser.py]: i've change it without another function. good call

File: sources/Tools/Visualiser.py (hunk 8 above)
Comment (lotoussa) [reply sources/Tools/Visualiser.py]: i've try something different, take a look

File: sources/Tools/Visualiser.py (hunk 10 above)
Comment (lotoussa) [reply sources/Tools/Visualiser.py]: i've remastered it, you can check

File: sources/Tools/DatasetHandler.py (hunk 4 above)
Comment (xaviave) [reply sources/Tools/DatasetHandler.py]: it s a public tool, it should be very generic, the only reason it's not a staticmethod it s because _load_npy is call, please change it

File: sources/Tools/Visualiser.py [hunk 14]
```diff
@@ -199,7 +223,10 @@ def _exit(exception=None, message="Error", mod=-1):
     def __init__(self, func=_scatter_plot_visualizer):
         # could add different matplotlib backend | for now to much work
         super().__init__()
-        self.visualizer_func = self.get_args("type_visualizer", default_value=func)
+        # check this setter for every program usable to prevent crash
+        [[self.header_visualizer, self.func_visualizer]] = self.get_args(
+            "type_visualizer", default_value={"advanced": func}
+        ).items()
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: items() is a looping method, i'm sure it works  but at what cost?

File: sources/Tools/Visualiser.py [hunk 15]
```diff
@@ -126,26 +125,51 @@ def _process_bar_data(raw_data, head):
...
+        plt.xlabel(xlabel)
+        plt.ylabel(ylabel)
+        plt.yticks([] if yticks is None else yticks)
+        plt.close()
         return fig
 
+    def _histogram_visualizer(self, header):
+        figures = [self._text_page()]
+        for feature in header:
+            try:
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: when could it crash, i think the try except is mostly used to catch ONE excpetions here is like a mega trash for the 2 funcs maybe rethinks it or call me for a debate

File: sources/Tools/Visualiser.py [hunk 16]
```diff
@@ -126,26 +125,51 @@ def _process_bar_data(raw_data, head):
...
-        rects2 = ax.bar(x + width / 2, hands["left"], width, label="Left")
-        ax.set_xticks(x)
-        ax.set_xticklabels(houses)
-        ax.legend()
-        self._auto_label(ax, rects1)
-        self._auto_label(ax, rects2)
-        fig.tight_layout()
+    @staticmethod
+    def _pick_by_elements(raw_data, filters, column_filter, feature):
+        return [raw_data.loc[raw_data[column_filter] == element, feature] for element in filters]
```
Comment (xaviave) [inline sources/Tools/Visualiser.py]: check if numpy.where could be useful

Comment (lotoussa) [reply sources/Tools/Visualiser.py]: i've already try it with pandas.query() but this is the most handle solution since we need to pick element from filter IN feature and we can't cut the data before to do it since we need to compare and check all theses terms at the same time.

**Review by xaviave — APPROVED** [review]
